core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -18,107 +18,96 @@ import (
-	CleanUpInterval = 15 * time.Minute
-	// ChannelSize represents the channel size for head broadcaster
-	ChannelSize = 20
+	// cleanUpInterval is the interval for cleaning up block maps
+	cleanUpInterval = 15 * time.Minute
+	// channelSize represents the channel size for head broadcaster
+	channelSize = 20
 )
 
-type BlockKey struct {
+type blockKey struct {
 	block int64
-	hash  common.Hash
+	hash  string
 }
 
-func (bk *BlockKey) getBlockKey() ocr2keepers.BlockKey {
+func (bk *blockKey) getBlockKey() ocr2keepers.BlockKey {
 	return ocr2keepers.BlockKey{
 		Number: ocr2keepers.BlockNumber(bk.block),
-		Hash:   bk.hash,
+		Hash:   common.HexToHash(bk.hash),
 	}
 }
 
 type BlockSubscriber struct {
-	sync                  utils.StartStopOnce
-	mu                    sync.RWMutex
-	ctx                   context.Context
-	cancel                context.CancelFunc
-	hb                    httypes.HeadBroadcaster
-	lp                    logpoller.LogPoller
-	headC                 chan BlockKey
-	unsubscribe           func()
-	subscribers           map[int]chan ocr2keepers.BlockHistory
-	blocksFromPoller      map[int64]common.Hash
-	blocksFromBroadcaster map[int64]common.Hash
-	maxSubId              int
-	lastClearedBlock      int64
-	lastSentBlock         int64
-	blockHistorySize      int64
-	lggr                  logger.Logger
+	sync             utils.StartStopOnce
+	mu               sync.RWMutex
+	ctx              context.Context
+	cancel           context.CancelFunc
+	hb               httypes.HeadBroadcaster
+	lp               logpoller.LogPoller
+	headC            chan *evmtypes.Head
+	unsubscribe      func()
+	subscribers      map[int]chan ocr2keepers.BlockHistory
+	blocks           map[int64]string
+	maxSubId         int
+	lastClearedBlock int64
+	lastSentBlock    int64
+	blockHistorySize int64
+	lggr             logger.Logger
 }
 
 func NewBlockSubscriber(hb httypes.HeadBroadcaster, lp logpoller.LogPoller, blockHistorySize int64, lggr logger.Logger) *BlockSubscriber {
 	return &BlockSubscriber{
-		hb:                    hb,
-		lp:                    lp,
-		headC:                 make(chan BlockKey, ChannelSize),
-		subscribers:           map[int]chan ocr2keepers.BlockHistory{},
-		blocksFromPoller:      map[int64]common.Hash{},
-		blocksFromBroadcaster: map[int64]common.Hash{},
-		blockHistorySize:      blockHistorySize,
-		lggr:                  lggr.Named("BlockSubscriber"),
+		hb:               hb,
+		lp:               lp,
+		headC:            make(chan *evmtypes.Head, channelSize),
+		subscribers:      map[int]chan ocr2keepers.BlockHistory{},
+		blocks:           map[int64]string{},
+		blockHistorySize: blockHistorySize,
+		lggr:             lggr.Named("BlockSubscriber"),
 	}
 }
 
-func (hw *BlockSubscriber) getBlockRange(ctx context.Context) ([]uint64, error) {
-	h, err := hw.lp.LatestBlock(pg.WithParentCtx(ctx))
+func (hw *BlockSubscriber) getBlockRange() ([]uint64, error) {
+	h, err := hw.lp.LatestBlock(pg.WithParentCtx(hw.ctx))
```

Comment:
do we need the log poller dependency here? thought we were trying to remove it

Reply:
the log poller is only used when this service starts. the reason is i am concerned that when this service just starts, if subscriber needs block history for 128 or 256 blocks, the head broadcaster's depth can only trace back to finality depth.
if this is not a serious issue, we can drop log poller dependency completely.